orttraining/orttraining/core/optimizer/batchnorm_replacement.cc

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "orttraining/core/optimizer/batchnorm_replacement.h"
+
+#include "core/common/logging/logging.h"
+#include "core/optimizer/rewrite_rule.h"
+#include "core/optimizer/utils.h"
+#include "core/graph/graph.h"
+#include "core/graph/graph_utils.h"
+
+namespace onnxruntime {
+
+Status BatchNormReplacement::Apply(Graph& graph, Node& bn_node, RewriteRuleEffect& rule_effect, const logging::Logger&) const {
+  const auto& bn_inputs = bn_node.MutableInputDefs();
+  auto& bn_outputs = bn_node.MutableOutputDefs();
+  const NodeArg* scale_input_def = bn_inputs[1];
+  auto scale_input_def_type_proto = scale_input_def->TypeAsProto();
+
+  // Guard against a BatchNorm that already has optional outputs present for some reason
+  if (bn_outputs.size() == 1) {
```

Comment:
what about the case bn_outputs.size() == 3?

Reply:
Fixed, thanks for catching this!